File: Python/Binary Tree - DFS/1372. Longest ZigZag Path in a Binary Tree.py
""" 1372. Longest ZigZag Path in a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/submissions/1388941787/ """

class Solution(object):
    def longestZigZag(self, root):
        ans = 0
        stack = [(root, 0, None)]
        while stack:
            node, n, left = stack.pop()
            if node:
                ans = max(ans, n)
                stack.append((node.left, 1 if left else n + 1, 1))
                stack.append((node.right, n + 1 if left else 1, 0))
        return ans

File: JavaScript/Binary Tree - DFS/1372. Longest ZigZag Path in a Binary Tree.js
/* 1372. Longest ZigZag Path in a Binary Tree
Difficulty: Medium
Submission Link: https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/submissions/1388941174/
*/

var longestZigZag = function(root) {
    let ans = 0;
    const stack = [[root, 0, null]];
    while (stack.length > 0) {
        const [node, n, left] = stack.pop();
        if (node) {
            ans = Math.max(ans, n);
            stack.push([node.left, left != null && left == 0 ? n + 1 : 1, 1]);
            stack.push([node.right, left != null && left == 1 ? n + 1 : 1, 0]);
        }
    }
    return ans;
};